apsuite/loco/utils.py

```diff
@@ -243,110 +244,148 @@ def jloco_calc_linear(config, matrix):
-            dip_kvalues = LOCOUtils.get_quads_strengths(
-                model, dip_indices)
-            set_quad_kdelta = LOCOUtils.set_dipset_kdelta
         else:
             dip_indices = config.respm.fam_data['BN']['index']
-            dip_kvalues = _np.array(
-                _pyaccel.lattice.get_attribute(model, 'KL', dip_indices))
-            set_quad_kdelta = LOCOUtils.set_dipmag_kdelta
-        matrix_nominal = LOCOUtils.respm_calc(
-            model, config.respm, config.use_dispersion)
 
-        dip_kmatrix = _np.zeros((
-            matrix_nominal.shape[0]*matrix_nominal.shape[1], len(dip_indices)))
+        magtype = 'dipole'
+        if config.parallel is not None:
+            dip_matrix = LOCOUtils._parallel_base(
+                config, model, dip_indices,
+                LOCOUtils._jloco_calc_k_matrix, magtype)
+        else:
+            dip_matrix = LOCOUtils._jloco_calc_k_matrix(
+                config, model, dip_indices, magtype)
+        return dip_matrix
 
-        model_this = _dcopy(model)
-        for idx, idx_set in enumerate(dip_indices):
-            set_quad_kdelta(
-                model_this, idx_set,
-                dip_kvalues[idx], config.DEFAULT_DELTA_K)
-            matrix_this = LOCOUtils.respm_calc(
-                model_this, config.respm, config.use_dispersion)
-            dmatrix = (matrix_this - matrix_nominal)/config.DEFAULT_DELTA_K
-            dip_kmatrix[:, idx] = dmatrix.ravel()
-            set_quad_kdelta(model_this, idx_set, dip_kvalues[idx], 0)
-        return dip_kmatrix
+    @staticmethod
+    def jloco_calc_k_quad(config, model):
+        """."""
+        if config.use_quad_families:
+            kindices = []
+            for fam_name in config.famname_quadset:
+                kindices.append(config.respm.fam_data[fam_name]['index'])
+        else:
+            kindices = config.respm.fam_data['QN']['index']
+        magtype = 'quadrupole'
+        if config.parallel is not None:
+            kmatrix = LOCOUtils._parallel_base(
+                config, model, kindices,
+                LOCOUtils._jloco_calc_k_matrix, magtype)
+        else:
+            kmatrix = LOCOUtils._jloco_calc_k_matrix(
+                config, model, kindices, magtype)
+        return kmatrix
 
     @staticmethod
-    def jloco_calc_k_sextupoles(config, model):
+    def jloco_calc_k_sext(config, model):
         """."""
-        sn_indices = config.respm.fam_data['SN']['index']
-        sn_kvalues = _np.array(
-            _pyaccel.lattice.get_attribute(model, 'KL', sn_indices))
-        set_quad_kdelta = LOCOUtils.set_quadmag_kdelta
+        if config.use_sext_families:
+            sindices = []
+            for fam_name in config.famname_sextset:
+                sindices.append(config.respm.fam_data[fam_name]['index'])
+        else:
+            sindices = config.respm.fam_data['SN']['index']
+        magtype = 'sextupole'
+        if config.parallel is not None:
```

Comment:
In order to cleanup the code you could put this if inside the parallel base method.

I also think the variable parallel could be False, True or an integer specifying the number of processes to use. In case it is True the code decides how many processes based on the number of cores of the machine.
For an example, take a look at my PR on pyaccel.